### PR TITLE
NEW RECORD TYPE: DNAME

### LINKED ISSUE #2880: Add DNAME support
DNAME records are used to delegate the whole DNS hierarchy to another zone.

Currently DNSControl does not support them but they are useful sometimes.

### PR DESCRIPTION
**Pull request by imlonghao**
close #2880

### CODE REVIEW

File: pkg/js/parse_tests/047-DNAME.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+D("foo.com","none",
+    DNAME("@", "bar.com")
```
Comment (tlimoncelli) [inline pkg/js/parse_tests/047-DNAME.js:2]:
```suggestion
    DNAME("@", "bar.com.")
```

File: pkg/js/parse_tests/047-DNAME.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  "domains": [
+    {
+      "name": "foo.com",
+      "registrar": "none",
+      "dnsProviders": {},
+      "records": [
+        {
+          "type": "DNAME",
+          "name": "@",
+          "target": "bar.com"
```
Comment (tlimoncelli) [inline pkg/js/parse_tests/047-DNAME.json:13]:
```suggestion
          "target": "bar.com."
```

File: integrationTest/integration_test.go [hunk 3]
```diff
@@ -1594,6 +1598,12 @@ func makeTests() []*TestGroup {
 			tc("Modify DHCID record", dhcid("test", "Test/AuCccgoJbsaxcQc9TUapptP69lOjxfNuVAA2kjEA=")),
 		),
 
+		testgroup("DNAME",
+			requires(providers.CanUseDNAME),
+			tc("Create DNAME record", dname("test", "example.com.")),
+			tc("Modify DNAME record", dname("test", "example.net.")),
```
Comment (tlimoncelli) [inline integrationTest/integration_test.go:1604]: Please also add a case where you add a non-FQDN name.  For example:

```
               testgroup("DNAME-short",
                       tc("Create a DNAME",
                               a("short", "1.2.3.4"),
                               dname("testdname", "short"),
                       ),
```

(I'm not sure if the `a()` is required)

**Review by tlimoncelli — APPROVED** [review]
Looks great! Thanks!
Just a few minor suggestions.

**Comment by imlonghao** [comment]
Test for cloudns

```
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/imlonghao.cloudns.asia
=== RUN   TestDNSProviders/imlonghao.cloudns.asia/Clean_Slate:Empty
=== RUN   TestDNSProviders/imlonghao.cloudns.asia/44:DNAME:Create_DNAME_record
    integration_test.go:236:
        + CREATE DNAME test.imlonghao.cloudns.asia example.com. ttl=300
=== RUN   TestDNSProviders/imlonghao.cloudns.asia/44:DNAME:Modify_DNAME_record
    integration_test.go:236:
        ± MODIFY DNAME test.imlonghao.cloudns.asia: (example.com. ttl=300) -> (example.net. ttl=300), ClouDNS ID: 482265345:
=== RUN   TestDNSProviders/imlonghao.cloudns.asia/44:DNAME:Create_DNAME_record_in_non-FQDN
    integration_test.go:236:
        - DELETE DNAME test.imlonghao.cloudns.asia example.net. ttl=300, ClouDNS ID: 482265345
    integration_test.go:236:
        + CREATE DNAME a.imlonghao.cloudns.asia b.imlonghao.cloudns.asia. ttl=300
--- PASS: TestDNSProviders (23.99s)
    --- PASS: TestDNSProviders/imlonghao.cloudns.asia (22.81s)
        --- PASS: TestDNSProviders/imlonghao.cloudns.asia/Clean_Slate:Empty (2.37s)
        --- PASS: TestDNSProviders/imlonghao.cloudns.asia/44:DNAME:Create_DNAME_record (5.98s)
        --- PASS: TestDNSProviders/imlonghao.cloudns.asia/44:DNAME:Modify_DNAME_record (6.11s)
        --- PASS: TestDNSProviders/imlonghao.cloudns.asia/44:DNAME:Create_DNAME_record_in_non-FQDN (7.17s)
=== RUN   TestDualProviders
    integration_test.go:337: Skipping.  DocDualHost == Cannot
--- SKIP: TestDualProviders (2.37s)
=== RUN   TestNameserverDots
    integration_test.go:401: Skipping.  DocDualHost == Cannot
--- SKIP: TestNameserverDots (2.38s)
PASS
ok  	github.com/StackExchange/dnscontrol/v4/integrationTest	29.165s
```

Test for bind

```
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/dnscontrol.net
=== RUN   TestDNSProviders/dnscontrol.net/Clean_Slate:Empty
File does not yet exist: "zones/dnscontrol.net.zone" (will create)
    integration_test.go:236:
        + CREATE dnscontrol.net SOA DEFAULT_NOT_SET. DEFAULT_NOT_SET. 3600 600 604800 1440 ttl=300
WRITING ZONEFILE: zones/dnscontrol.net.zone
=== RUN   TestDNSProviders/dnscontrol.net/44:DNAME:Create_DNAME_record
    integration_test.go:236:
        + CREATE test.dnscontrol.net DNAME example.com. ttl=300
WRITING ZONEFILE: zones/dnscontrol.net.zone
=== RUN   TestDNSProviders/dnscontrol.net/44:DNAME:Modify_DNAME_record
    integration_test.go:236:
        ± MODIFY test.dnscontrol.net DNAME (example.com. ttl=300) -> (example.net. ttl=300)
WRITING ZONEFILE: zones/dnscontrol.net.zone
=== RUN   TestDNSProviders/dnscontrol.net/44:DNAME:Create_DNAME_record_in_non-FQDN
    integration_test.go:236:
        + CREATE a.dnscontrol.net DNAME b.dnscontrol.net. ttl=300
        - DELETE test.dnscontrol.net DNAME example.net. ttl=300
WRITING ZONEFILE: zones/dnscontrol.net.zone
--- PASS: TestDNSProviders (0.01s)
    --- PASS: TestDNSProviders/dnscontrol.net (0.01s)
        --- PASS: TestDNSProviders/dnscontrol.net/Clean_Slate:Empty (0.00s)
        --- PASS: TestDNSProviders/dnscontrol.net/44:DNAME:Create_DNAME_record (0.00s)
        --- PASS: TestDNSProviders/dnscontrol.net/44:DNAME:Modify_DNAME_record (0.00s)
        --- PASS: TestDNSProviders/dnscontrol.net/44:DNAME:Create_DNAME_record_in_non-FQDN (0.00s)
=== RUN   TestDualProviders
    integration_test.go:358: Clearing everything
    integration_test.go:352: #1:
        - DELETE a.dnscontrol.net DNAME b.dnscontrol.net. ttl=300
WRITING ZONEFILE: zones/dnscontrol.net.zone
    integration_test.go:365: Adding nameservers from another provider
    integration_test.go:352: #1:
        + CREATE dnscontrol.net NS ns1.example.com. ttl=300
        + CREATE dnscontrol.net NS ns2.example.com. ttl=300
WRITING ZONEFILE: zones/dnscontrol.net.zone
    integration_test.go:368: Running again to ensure stability
--- PASS: TestDualProviders (0.00s)
=== RUN   TestNameserverDots
=== RUN   TestNameserverDots/No_trailing_dot_in_nameserver
--- PASS: TestNameserverDots (0.00s)
    --- PASS: TestNameserverDots/No_trailing_dot_in_nameserver (0.00s)
PASS
ok  	github.com/StackExchange/dnscontrol/v4/integrationTest	0.384s
```

**Comment by tlimoncelli** [comment]
Fantastic!

Thank you so much!   Many people have requested this feature!  I'm sure they'll be very happy!

**Pull request merged**